packages/kokkos-kernels/src/sparse/KokkosSparse_spadd.hpp

```diff
@@ -182,125 +203,126 @@ namespace Experimental {
     CcolindsT ABperm;
   };
 
-  template<typename size_type, typename KeyType, typename ValueType, typename IndexType>
-  KOKKOS_INLINE_FUNCTION void
-  radixSortKeysAndValues(KeyType* keys, KeyType* keysAux, ValueType* values, ValueType* valuesAux, IndexType n)
+  template<typename ExecSpace, typename size_type, typename ordinal_type, typename CrowptrsT, typename CcolindsT>
+  struct SortEntriesFunctor
   {
-    if(n <= 1)
-      return;
-    //sort 4 bits at a time
-    KeyType mask = 0xF;
-    bool inAux = false;
-    //maskPos counts the low bit index of mask (0, 4, 8, ...)
-    IndexType maskPos = 0;
-    IndexType sortBits = 0;
-    KeyType minKey = keys[0];
-    KeyType maxKey = keys[0];
-    for(size_type i = 0; i < n; i++)
-    {
-      if(keys[i] < minKey)
-        minKey = keys[i];
-      if(keys[i] > maxKey)
-        maxKey = keys[i];
-    }
-    //subtract a bias of minKey so that key range starts at 0
-    for(size_type i = 0; i < n; i++)
-    {
-      keys[i] -= minKey;
-    }
-    KeyType upperBound = maxKey - minKey;
-    while(upperBound)
-    {
-      upperBound >>= 1;
-      sortBits++;
-    }
-    for(size_type s = 0; s < (sortBits + 3) / 4; s++)
+    SortEntriesFunctor(const CrowptrsT& Crowptrs_, const CcolindsT& Ccolinds_, const CcolindsT& ABperm_) :
+      Crowptrs(Crowptrs_),
+      Ccolinds(Ccolinds_),
+      CcolindsAux("C colind aux", Ccolinds_.extent(0)),
+      ABperm(ABperm_),
+      ABpermAux("AB perm aux", ABperm_.extent(0))
+    {}
+
+    KOKKOS_INLINE_FUNCTION static void
+    radixSortKeysAndValues(ordinal_type* keys, ordinal_type* keysAux, ordinal_type* values, ordinal_type* valuesAux, size_type n)
```

Comment:
@brian-kelley @csiefer2 It looks like it wouldn't be too hard to convert this into a team-parallel sort.  I might do that at some point.  Do you see any obstacles to that?

Reply:
@mhoemmen It will be after the KokkosKernels snapshot. On GPU it will be parallel (team) bitonic sort within rows, and on CPU it's still serial radix sort within rows. This would also (probably) give a big speedup to Tpetra::sortCrsEntries().

Reply:
@mhoemmen If you meant parallelizing within the radix sort, I have thought about it and I don't think it's possible/easy without shared memory atomics, which are apparently no faster than global memory atomics.

Reply:
@brian-kelley Right now I have no team sort, so as long as a team sort is faster than a sequential sort, I would take it :-) .